### ISSUE TITLE
Inline enum values should be cast to their enum type just like inline bind variables

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#5612]
See also:
http://www.jooq.org/doc/latest/manual/sql-execution/stored-procedures/#comment-2961018657

**Comment by jottyfan** [comment]
seems not only to affect stored procedures, but also the values part of sql:

````sql
create type enum_myenum as enum('this', 'is', 'crap');
create table t_mytable(e enum_myenum);
\d t_mytable
    Tabelle »public.t_mytable«
 Spalte |     Typ     | Attribute 
--------+-------------+-----------
 e      | enum_myenum | 
````

Using jooq 3.8.6, the insert statement generated  by that code
````java
create.insertInto(T_MYTABLE, T_MYTABLE.E).values(EnumMyenum.this).execute();
````
generates that sql:
````sql
insert into "t_mytable"("e") values (?)
````
but should generate that sql for correct jdbc usage:
````sql
insert into "t_mytable"("e") values (?::"enum_myenum")
````

**Comment by jottyfan** [comment]
found a workaround myself, but this seems not to be that cool, but anyway, it works:
````java
create.execute("insert into t_mytable(e) values (?::enum_myenum)", EnumMyenum.this);
````

**Comment by triptec** [comment]
Is there a eta on this, currently using 3.9.1 and do not want to use the workaround suggested here..

**Comment by ben-manes** [comment]
Ran into this too, would be nice if enums were casted.

**Comment by ben-manes** [comment]
It appears that `AbstractBinding#shouldCast` always returns false, with the claim `Generated enums should not be cast...`. I'm not sure why that would make sense.

**Comment by parbhatia** [comment]
Recently came across this issue in JOOQ v3.16.4 using Postgres.
This issue prevents the use of convenient inserts/updates using POJOs that use enum types.

**Comment by parbhatia** [comment]
> Recently came across this issue in JOOQ v3.16.4 using Postgres. This issue prevents the use of convenient inserts/updates using POJOs that use enum types.

Nevermind. Was not using the correct SQL Dialect. This issue is not present in Postgres 14 on JOOQ v3.16.5

**Comment by lukaseder** [comment]
Thanks for your various comments, folks. While I think that most comments are related to a wrong SQL dialect having been used (because enum *bind values* have been cast to their enum type for a long time now. E.g. there's been a related task in jOOQ 3.6: https://github.com/jOOQ/jOOQ/issues/3778).

However, this is about casting enum *inline values*, e.g. when using `DSL::inline` or `StatementType.STATIC_STATEMENT`. In a lot of cases, PostgreSQL will implicitly cast the enum type to a `varchar` or the other way round, but there may be some cases where this doesn't work, e.g. when calling overloaded stored functions.

So, I'm fixing this now. The fix will still assume the correct configuration of the SQL dialect, of course.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.6 (https://github.com/jOOQ/jOOQ/issues/13353), and 3.15.10 (https://github.com/jOOQ/jOOQ/issues/13354)
